genie-web/src/main/java/com/netflix/genie/web/agent/services/impl/AgentConfigurationServiceImpl.java

```diff
@@ -0,0 +1,138 @@
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Sets;
+import com.netflix.genie.web.agent.services.AgentConfigurationService;
+import com.netflix.genie.web.properties.AgentConfigurationProperties;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.EnumerablePropertySource;
+import org.springframework.core.env.Environment;
+import org.springframework.core.env.PropertySource;
+import org.springframework.core.env.PropertySources;
+import org.springframework.core.env.StandardEnvironment;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+/**
+ * This implementation of {@link AgentConfigurationService} forwards properties set on the server that match a given
+ * set of regular expressions, plus any additional ones specified in configuration.
+ * It utilizes a cache to avoid recomputing the set of properties and values for every request.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Slf4j
+public class AgentConfigurationServiceImpl implements AgentConfigurationService {
+
+    private static final String AGENT_PROPERTIES_CACHE_KEY = "agent-properties";
+
+    private final AgentConfigurationProperties agentConfigurationProperties;
+    private final Environment environment;
+    private final Pattern agentPropertiesPattern;
+    private final LoadingCache<String, Map<String, String>> cache;
+
+    /**
+     * Constructor.
+     *
+     * @param agentConfigurationProperties the properties
+     * @param environment the environment
+     */
+    public AgentConfigurationServiceImpl(
+        final AgentConfigurationProperties agentConfigurationProperties,
+        final Environment environment
+    ) {
+        this.agentConfigurationProperties = agentConfigurationProperties;
+        this.environment = environment;
+
+        this.agentPropertiesPattern = Pattern.compile(
+            this.agentConfigurationProperties.getAgentPropertiesFilterPattern(),
+            Pattern.CASE_INSENSITIVE
+        );
+
+        // Use a cache to re-compute agent properties periodically, rather than for every request.
+        this.cache = Caffeine
+            .newBuilder()
+            .expireAfterWrite(this.agentConfigurationProperties.getCacheExpirationInterval())
+            .refreshAfterWrite(this.agentConfigurationProperties.getCacheRefreshInterval())
+            .initialCapacity(1)
+            .build(this::loadProperties);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Map<String, String> getAgentProperties() {
+        return this.cache.get(AGENT_PROPERTIES_CACHE_KEY);
+    }
+
+
+    private Map<String, String> loadProperties(@NonNull final String propertiesKey) {
+        if (AGENT_PROPERTIES_CACHE_KEY.equals(propertiesKey)) {
```

Comment:
Why so many levels of indirection here between the getter above and the load function in the cache and the reload properties below?

Reply:
I suppose i could inline the function below here.